proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,330 @@
+  that depends on the implementation being well engineered etc (GHC team);
+  and the implementor should make an explicit proposal to the CLC specifying the precise changes.
+
+Abstraction leakage
+-------------------
+
+We may foresee a couple of ways in which changes in ``ghc-internals`` could become client visible:
+
+- Occasionally, an error message may mention a fully qualified name for an out-of-scope identifier.
+  For example (GHC test ``mod153``)::
+
+    Ambiguous occurrence ‘id’
+               It could refer to either ‘Prelude.id’,
+                            imported from ‘Prelude’ at mod153.hs:2:8
+                            (and originally defined in ‘GHC.Base’)
+                         or ‘M.id’, defined at mod153.hs:2:21
+
+  The "originally defined in" mentions a module; and if that module is in a package that is not imported, GHC will package-qualify the module name.
+  And seeing ``ghc-internals:GHC.Base`` is perhaps less nice.
+  This is not a new problem: we already package-qualify modules in ``ghc-prim``.
+  One solution is to remove the "originally defined in.." parenthesis for types and functions that would require such package qualification.
+
+- Another form of leakage could be: a new class in ``ghc-internals``, *not exposed in base*, that is given instances for existing data types.
+  There is a risk that those instances might confusingly be visible to clients of ``base``.
+  If so, the CLC should at least be consulted.
+
+These issues concern error messages and documentation, neither of which are in the direct scope of CLC.
+They are not new because we already have ``ghc-prim``.
+They may not be show-stoppers, but we should be thoughtful about mitigating them.
+
+Versions and backports
+----------------------
+
+We agree that the version number of ``ghc-internals`` may have a major bump between minor releases of GHC.
+(Why? Because to fix the bug we change something in ``ghc-internals``.)
+
+This makes an exception to a general rule: generally, a minor release of GHC (say 9.6.4) which only fixes bugs, never makes a major version bump to ``base``, or indeed any boot package.
+
+We should discuss this (rather important) exception with the Stackage curators.
+
+But this same issue could in principle affect ``base`` too.
+Very occasionally a **bug-fix** might involve a change to the user-visible API.
+Example: `role annotations on SNat <https://github.com/haskell/core-libraries-committee/issues/170>`__ (although there is a debate as to whether this specific change constitutes a "breaking change" under the PVP).
+
+Under these circumstances we (together) will have to decide whether to
+
+- Back port the fix, and not bump the major version of ``base`` (i.e. bend the PVP), or
+- Bump the major version of base, but therefore be unable to fix the bug in the released GHC.
+
+This is a decision for the CLC.
+See PVP issue https://github.com/haskell/pvp/issues/10.
```

Comment:
This is a technical limitation of the way that `base` and GHC work, and would in principle be fixed by allow a GHC version to work with multiple versions of `base`, right?

Reply:
Are you proposing (eventually) to continue to ship the `base` with the bug and a new one with the bug? That would work. But I think the point in this case was that some bugs are not a matter of implementation but inherent to the interface itself.

(I remain very pro multiple `base` versions with one GHC, to be clear :).)

Reply:
I think I'm confused - can a released GHC use a different version of `base`? I was under the impression that the `base <-> GHC` relationship was fixed, and every version of GHC can only use a single version of `base`. That means that GHC X.Y.Z must always use the same `base` version, and a new `base` version means a new release of GHC.

Reply:
I think there is much confusion around this topic because people confuse GHC the software with GHC the official bindist.

The answer is Yes, and it's not even a new feature. See https://gitlab.haskell.org/ghc/ghc/-/blob/master/compiler/GHC/Unit/Types.hs#L562.

The short version that is one is liable to shoot themselves in the foot if they have more than one `base` in a package database, but one is always free to have multiple package databases they will not mix together. In each "parallel universe" of the separate package database, there is only one `base`, and no potential for confusion.

GHC in fact has *no way of knowing* whether this is the "official base" that is was shipped with the bindist, or ones own. How could it now? (It isn't shipped with trusted ABI hashes for `base`, for example, on the contrary, it expects `base` to have *no* ABI hashes; it is maximally trusting of whatever `base` you give it!)

For a variety of reasons, no one has yet proposed we take advantage of this. But we can. There is no technical limitation, just the ergonomic/human problem of it being easy to get our package databases with their separate `base`s mixed up.

(I believe @alt-romes might be working on getting all wired-ing packages ABI hashes to solve the humans-getting-confused problem, but I am not sure about that.)

Reply:
> (I remain very pro multiple base versions with one GHC, to be clear :).)

I think we all are in favour of that.   

But, to be clear, achieving that goal is not part of the current document.  So this thread is somewhat "by the way".

Still, the plan outlined this document should make the goal of a reinstallable base much more achievable.  GHC knows (wired into its binary) where, say `GHC.Base.map` is defined.  If we move that definition from one module to another, everything will stop working.  I think that is one reason that `base` and GHC are so tightly bound together.  But if  `GHC.Base.map` was in `ghc-internals` then GHC and `ghc-internals` would remain tightly bound together, but `base` would be much more loosely coupled.  I don't yet see any technical obstacle to installing a fresh `base` without changing `ghc`.

I could be wrong; but all that matters for *this* conversation is that the plan here is a step in the right direction for the reinstallable-base goal.

Reply:
Fantastic, thanks for the clarification!

Allowing `base` to iterate without incurring a GHC release cycle would be a big win for the ecosystem, for sure.